src/ucp/api/cuda/ucp.cuh

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2025. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCP_CUH
+#define UCP_CUH
+
+#include <ucs/type/status.h>
+#include <uct/api/uct.h>
+#include <ucp/api/ucp.h>
+
+
+/* TODO: Use from ucp.h */
+typedef struct ucp_dlist_elem {} ucp_dlist_elem_t;
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Descriptor list handle stored on GPU memory.
+ *
+ * This handle is obtained and managed with functions called on host. It can be
+ * used repeatedly from GPU code to perform memory transfers.
+ *
+ * The handle and most of its content is stored on GPU memory, with the intent
+ * to be as memory-local as possible.
+ */
+typedef struct {
+    /**
+     * Host context to be used by host handle management functions, stored in
+     * host memory.
+     */
+    void             *host_context;
+
+    /**
+     * Protocol index computed by host handle management functions when
+     * creating handle.
+     */
+    int              proto_idx;
+
+    /**
+     * Array of pointers to UCT exported endpoints, used for multi-lane
+     * transfers.
+     */
+    uct_ep_t         **uct_ep;
+
+    /**
+     * Number of UCT exported endpoints found in @ref uct_ep arrays.
+     */
+    unsigned         num_uct_eps;
+
+    /**
+     * Number of entries in the descriptor list array @ref elems.
+     */
+    size_t           dlist_length;
+
+    /**
+     * Array of descriptor list containing memory pairs to be used by GPU
+     * device functions for memory transfers.
+     */
+    ucp_dlist_elem_t elems[];
+} ucp_dlist_handle_t;
+
+
+typedef ucp_dlist_handle_t *ucp_dlist_handle_h;
+
+
+/**
+ * @ingroup UCP_COMM
+ * @brief Posts one memory put operation.
+ *
+ * This GPU routine posts one put operation using descriptor list handle.
+ * The @ref dlist_index is used to point at the dlist entry to be used for the
+ * memory transfer. The addresses and length must be valid for the used dlist
+ * entry.
+ *
+ * This routine can be called repeatedly with the same handle and different
+ * addresses and length. The flags parameter can be used to modify the behavior
+ * of the routine.
+ *
+ * @param [in]  handle      Exported descriptor list handle to use for transfer.
+ * @param [in]  addr        Local virtual address to send data from.
+ * @param [in]  remote_addr Remote virtual address to send data to.
+ * @param [in]  length      Length in bytes of the data to send.
+ * @param [in]  dlist_index Index in descriptor list pointing to the memory
+ *                          registration keys to use for the transfer.
+ * @param [in]  flags       Bitfield usable to modify the function behavior.
+ * @param [out] status      GPU status returned by the call.
+ */
+__device__ void
+ucp_gpu_post_single(ucp_dlist_handle_h handle,
```

Comment:
It might be better to align it with the existing UCP API.
We have non-blocking remote memory put operation `ucp_put_nbx`, so I'd suggest `ucp_gpu_put`. I think we don't need to add `_nb` (non-blocking), or  `_nbx` (+scalable). Because originally it was just `ucp_put`. And the postfixes were added to introduce new versions of put operation.

We also have `ucp_atomic_op_nbx` to post an atomic. So I'd suggest `ucp_gpu_atomic`. Not sure it we need to make it `ucp_gpu_atomic_op`, along with introducing `ucp_gpu_atomic_op_t`, so we will be able to add other operation types.

`ucp_gpu_post_multi` -> `ucp_gpu_put_multi`